### PR TITLE
BUG: crosstab with duplicate column or index labels

### LINKED ISSUE #22529: Crosstab Not Working with Duplicate Column Labels
#### Code Sample, a copy-pastable example if possible

# 1. Example with 2 lambdas applied on same column
df = pd.DataFrame([[0, 1], [0, 1], [1, 0]], columns=['a', 'b'])
df
   a b
0 0 1
1 0 1
2 1 0

pd.crosstab(df['a'].apply(lambda x: x), df['a'].apply(lambda x: x+1))
a 1 2
a
1 2 0
2 0 1

# 2. Example by manually renaming a column
df = pd.DataFrame([[0, 1], [0, 1], [1, 0]], columns=['a', 'b'])
df
   a b
0 0 1
1 0 1
2 1 0

pd.crosstab(df['a'], df['b'])
b 0 1
a
0 0 2
1 1 0

s = df['b']
s.name = 'a'
pd.crosstab(df['a'], s)
a 0 1
a
0 1 0
1 0 2

In both cases, the output is not the one expected. The crosstab applies one Series to itself.

#### Problem description

I encountered the problem when using a crosstab on the same column with 2 different lambdas. The crosstab output a crosstab applying the second column to itself. It seems that crosstab is confused by the same name of the 2 series.

The problem is the same when the 2nd column is manually renamed with the name of the 1rst one, before applying the crosstab. See examples.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.2
pip: 10.0.0
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #22529
- [x] tests added / passed
- [x] passes black pandas
- [x] passes git diff upstream/master -u -- "*.py" | flake8 --diff
- [x] whatsnew entry

Picking up from #28474 

cc @jreback in case this can go in in time for 1.2

### CODE REVIEW

File: pandas/tests/reshape/test_crosstab.py [hunk 1]
```diff
@@ -535,17 +535,51 @@ def test_crosstab_with_numpy_size(self):
...
-            np.eye(3, dtype=np.int64), index=expected_index, columns=expected_index
+        # Same name in both rows and columns
+        s1 = Series(range(3), name="foo")
+        s2 = s1 + 1
+        expected = crosstab(s1, s2.rename("bar")).rename_axis(
+            columns={"bar": "foo"}, axis=1
+        )
+        result = crosstab(s1, s2)
+        tm.assert_frame_equal(result, expected)
+        assert result.index.names == ["foo"]
```
Comment (arw2019) [inline pandas/tests/reshape/test_crosstab.py:550]: getting rid of these asserts since `assert_frame_equal` already does this

**Comment by arw2019** [comment]
I tried consolidating the tests using `eval` but for a reason I can't atm figure out this doesn't work:
``` python
    @pytest.mark.parametrize("args_res, args_exp, axis", (
        ["s3, [s1, s2_foo]", "s3, [s1, s2_bar]", 1],
        ["[s1, s2_foo], s3", "[s1, s2_bar], s3", 0],
    ))
    def test_crosstab_duplicate_labels(self, args_res, args_exp, axis):
        # GH 22529

        s1 = Series(range(3), name="foo")
        s2_foo = Series(range(1, 4), name="foo")
        s2_bar = Series(range(1, 4), name="bar")
        s3 = Series(range(3), name="waldo")

        result = crosstab(*eval(args_res))

        expected = crosstab(*eval(args_exp))
        expected = expected.rename_axis(columns={"bar": "foo"}, axis=axis)

        tm.assert_frame_equal(result, expected)

```

File: pandas/tests/reshape/test_crosstab.py [hunk 2]
```diff
@@ -535,15 +535,32 @@ def test_crosstab_with_numpy_size(self):
         )
         tm.assert_frame_equal(result, expected)
 
-    def test_crosstab_dup_index_names(self):
-        # GH 13279
-        s = Series(range(3), name="foo")
+    def test_crosstab_duplicate_names(self):
```
Comment (arw2019) [inline pandas/tests/reshape/test_crosstab.py:538]: addressed https://github.com/pandas-dev/pandas/pull/28474/files#r432993578

File: pandas/core/reshape/pivot.py [hunk 3]
```diff
@@ -731,3 +746,57 @@ def _get_names(arrs, names, prefix: str = "row"):
...
+
+
+def _build_names_mapper(
+    names: List[str], shared_col_row_names: Set[str], suffix: str
+) -> Tuple[Dict[str, str], List[str]]:
+    """
+    Given a list of row or column names, creates a mapper of unique names to
+    column/row names.
+
+    Parameters
```
Comment (arw2019) [inline pandas/core/reshape/pivot.py:758]: working on https://github.com/pandas-dev/pandas/pull/28474/files#r432993840

File: pandas/core/reshape/pivot.py [hunk 4]
```diff
@@ -578,29 +580,39 @@ def crosstab(
...
-    if values is None and aggfunc is not None:
-        raise ValueError("aggfunc cannot be used without values.")
+    rownames = _get_names(index, rownames, prefix="row")
+    colnames = _get_names(columns, colnames, prefix="col")
 
-    if values is not None and aggfunc is None:
-        raise ValueError("values cannot be used without an aggfunc.")
+    # We create our own mapping of row and columns names
+    # to prevent issues with duplicate columns/row names. GH Issue: #22529
+    shared_col_row_names = set(rownames).intersection(set(colnames))
```
Comment (arw2019) [inline pandas/core/reshape/pivot.py:602]: working on https://github.com/pandas-dev/pandas/pull/28474/files#r491538724

Comment (arw2019) [reply pandas/core/reshape/pivot.py]: addressed

File: pandas/core/reshape/pivot.py (hunk 3 above)
Comment (arw2019) [reply pandas/core/reshape/pivot.py]: addressed

File: pandas/core/reshape/pivot.py [hunk 5]
```diff
@@ -731,3 +743,30 @@ def _get_names(arrs, names, prefix: str = "row"):
             names = list(names)
 
     return names
+
+
+def _build_names_mapper(
+    rownames: List[str], colnames: List[str]
+) -> Tuple[Dict[str, str], List[str], Dict[str, str], List[str]]:
+    def get_duplicates(names):
```
Comment (jreback) [inline pandas/core/reshape/pivot.py:751]: can you add a doc-string here

**Review by jreback — CHANGES_REQUESTED** [review]
looks good thanks for picking this up. comments & ping on green.

File: pandas/core/reshape/pivot.py (hunk 5 above)
Comment (arw2019) [reply pandas/core/reshape/pivot.py:778]: Done

**Comment by arw2019** [comment]
Green + addressed comments

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019 and @cuchoi

**Pull request merged**